### ISSUE TITLE
Can't locate object method "new" via package "JSON::XS"

### ISSUE THREAD

**Issue by aspartam1987** [linuxserver/docker-tvheadend#83]
Hello guys,

I installed the tvheadend docker on unRAID 6.5.0. When I wanted to configure XMLTV grabber I got an error. The command I used was:

`docker exec -it tvheadend /usr/bin/tv_grab_huro --configure`

The output is:

> using config filename /config/.xmltv/tv_grab_huro.conf
> Grab listings for which country?
> 0: Czech
> 1: Hungary
> 2: Romania
> 3: Slovakia
> Select one: [0,1,2,3 (default=1)] 1
> getting list of channels: Can't locate object method "new" via package "JSON::XS" (perhaps you forgot to load "JSON::XS"?) at/usr/bin/tv_grab_huro line 1239, <STDIN> line 1.

Then I thought maybe there is a problem in the "/usr/bin/tv_grab_huro" script, so I edited line 137 from "use JSON" to "use JSON::XS". The reason was the line 1239, where stands this:

> my $json_data = (($DEFAULT_ENCODING !~ /utf\-?8/i) && ($COUNTRY eq 'hu')) ? JSON::XS->new->utf8(0)->decode(encode($DEFAULT_ENCODING,decode('utf-8', $data))) : JSON::XS->new->utf8(0)->decode($data) or`

and tried to configure the XMTLV grabber again.
The output was:

> Can't locate JSON/XS.pm in @ INC (you may need to install the JSON::XS module) (@ INC contains: /usr/local/lib/perl5/site_perl /usr/local/share/perl5/site_perl /usr/lib/perl5/vendor_perl /usr/share/perl5/vendor_perl /usr/lib/perl5/core_perl /usr/share/perl5/core_perl) at /usr/bin/tv_grab_huro line 137.
> BEGIN failed--compilation aborted at /usr/bin/tv_grab_huro line 137.
> 

So I tried to install the module with command:

`docker exec -it tvheadend perl -MCPAN -e 'install JSON::XS'`

The messages that I get:

> CPAN.pm requires configuration, but most of it can be done automatically.
> If you answer 'no' below, you will enter an interactive dialog for each
> configuration option instead.
> 
> Would you like to configure as much as possible automatically? [yes]
> 
> 
> ALERT: 'make' is an essential tool for building perl Modules.
> Please make sure you have 'make' (or some equivalent) working.
> 
> 
> Autoconfiguration complete.
> 
> commit: wrote '/config/.cpan/CPAN/MyConfig.pm'
> 
> You can re-run configuration any time with 'o conf init' in the CPAN shell
> Fetching with LWP:
> http://www.cpan.org/authors/01mailrc.txt.gz
> Reading '/config/.cpan/sources/authors/01mailrc.txt.gz'
> ............................................................................DONE
> Fetching with LWP:
> http://www.cpan.org/modules/02packages.details.txt.gz
> Reading '/config/.cpan/sources/modules/02packages.details.txt.gz'
>   Database was generated on Wed, 18 Apr 2018 09:54:33 GMT
> ............................................................................DONE
> Fetching with LWP:
> http://www.cpan.org/modules/03modlist.data.gz
> Reading '/config/.cpan/sources/modules/03modlist.data.gz'
> DONE
> Writing /config/.cpan/Metadata
> Running install for module 'JSON::XS'
> Fetching with LWP:
> http://www.cpan.org/authors/id/M/ML/MLEHMANN/JSON-XS-3.04.tar.gz
> Fetching with LWP:
> http://www.cpan.org/authors/id/M/ML/MLEHMANN/CHECKSUMS
> Checksum for /config/.cpan/sources/authors/id/M/ML/MLEHMANN/JSON-XS-3.04.tar.gz ok
> 'YAML' not installed, will not store persistent state
> ---- Unsatisfied dependencies detected during ----
> ----       MLEHMANN/JSON-XS-3.04.tar.gz       ----
>     Canary::Stability [build_requires]
> Running install for module 'Canary::Stability'
> Fetching with LWP:
> http://www.cpan.org/authors/id/M/ML/MLEHMANN/Canary-Stability-2012.tar.gz
> Checksum for /config/.cpan/sources/authors/id/M/ML/MLEHMANN/Canary-Stability-2012.tar.gz ok
> Configuring M/ML/MLEHMANN/Canary-Stability-2012.tar.gz with Makefile.PL
> Checking if your kit is complete...
> Looks good
> Generating a Unix-style Makefile
> Writing Makefile for Canary::Stability
> Writing MYMETA.yml and MYMETA.json
>   MLEHMANN/Canary-Stability-2012.tar.gz
>   /usr/bin/perl Makefile.PL -- OK
> Running make for M/ML/MLEHMANN/Canary-Stability-2012.tar.gz
>   MLEHMANN/Canary-Stability-2012.tar.gz
>   make -- NOT OK
>   No such file or directory
>   MLEHMANN/JSON-XS-3.04.tar.gz
>   Has already been unwrapped into directory /config/.cpan/build/JSON-XS-3.04-0
> Warning: Prerequisite 'Canary::Stability => 0' for 'MLEHMANN/JSON-XS-3.04.tar.gz' failed when processing 'MLEHMANN/Canary-Stability-2012.tar.gz' with 'make => NO'. Continuing, but chances to succeed are limited.
> Configuring M/ML/MLEHMANN/JSON-XS-3.04.tar.gz with Makefile.PL
> Can't locate Canary/Stability.pm in @ INC (you may need to install the Canary::Stability module) (@ INC contains: /usr/local/lib/perl5/site_perl /usr/local/share/perl5/site_perl /usr/lib/perl5/vendor_perl /usr/share/perl5/vendor_perl /usr/lib/perl5/core_perl /usr/share/perl5/core_perl .) at Makefile.PL line 4.
> BEGIN failed--compilation aborted at Makefile.PL line 4.
> Warning: No success on command[/usr/bin/perl Makefile.PL]
>   MLEHMANN/JSON-XS-3.04.tar.gz
>   /usr/bin/perl Makefile.PL -- NOT OK

So I found out that "make" is removed from docker, so I can't install modules.

I would like to ask you, to include JSON::XS module and if its necessary the Canary::Stability too

In summary:
Host OS: unRAID 6.5.0
Command what makes the problem: `docker exec -it tvheadend /usr/bin/tv_grab_huro --configure`
The messsage I get: 

> Can't locate object method "new" via package "JSON::XS" (perhaps you forgot to load "JSON::XS"?) at/usr/bin/tv_grab_huro line 1239, <STDIN> line 1.

The solution might be: adding module(s) JSON::XS (and Canary::Stability if necessary)

Thanks!
Daniel

**Comment by tobbenb** [comment]
I'll have a look at it in the weekend or before if I get the time.

**Comment by tobbenb** [comment]
It works here now adding the JSON::XS package. I'll push a fix soon.
When running the configuration, add -u abc after -it like below. It's mentioned in the readme and must be done so the permissions are correct for the abc user.

'docker exec -it -u abc tvheadend /usr/bin/tv_grab_huro --configure'
